feat(llm.txt): adds script to compile llm.txt

brookton opened the pull request:
# Add LLM.txt file generation script

## Addresses open discussions:
- [#7241](https://github.com/orgs/mantinedev/discussions/7241)
- [#7618](https://github.com/orgs/mantinedev/discussions/7618)
- [#7736](https://github.com/orgs/mantinedev/discussions/7736)
- [#7976](https://github.com/orgs/mantinedev/discussions/7976)

## Requirements
The script must:
- [x] compile mdx 
- [x] extract all demos 
- [x] combine with props tables and styles api data
- [x] include all faq content from help.mantine.dev

## Notes:
- can be added to pipelines and auto generate or update as new changes are made
- a frontend doc about llm txt stuff similar to [chakra-ui's llm docs](https://chakra-ui.com/docs/get-started/llms) might be worth considering as well

brookton commented:
> Does not work:
> 
> * No demos code
> * No props descriptions
> * No Styles API data
> 
> ```
> ### accordion
> File: apps/mantine.dev/src/pages/core/accordion.mdx
> 
> export default Layout(MDX_DATA.Accordion);
> 
> ## Usage
> 
> Accordion allows users to expand and collapse sections of content.
> It helps manage large amounts of information in a limited space
> by showing only the section headers initially and revealing content on interaction.
> 
> Accordion is commonly used for:
> 
> * FAQ sections: displaying questions as headers with answers revealed on click
> * Forms: organizing long forms into sections, for example, personal info, shipping, and payment
> * Menus: nested navigation in sidebars or mobile views
> 
> \[DEMO: AccordionDemos.configurator]
> 
> ## Change chevron
> 
> Use the `chevron` prop to change the chevron icon. When `chevron` is set,
> `chevronIconSize` prop is ignored. To remove the chevron icon, use `chevron={null}`.
> 
> To customize chevron styles, use [Styles API](/styles/styles-api/) with
> [data-rotate](/styles/data-attributes/) attribute. It is set when the item
> is opened if the `disableChevronRotation` prop is not set.
> 
> Example of a custom chevron icon with rotation styles:
> 
> \[DEMO: AccordionDemos.chevron]
> ```

I see what you mean here, definitely tweaking it a bit still.

rtivital commented:
Leftovers:
- `<AutoContrast component="Button" />`
- `<ComboboxData component="Autocomplete" />`
- `<ComboboxFiltering component="Autocomplete" />`
- `<ComboboxLargeData component="Autocomplete" />`
- `<FlexboxGapSupport component="Flex" />`
- `<GetElementRef component="Input" refType="input" />`
- `<Gradient component="Text" />`
- `<InputAccessibility component="Autocomplete" />`
- `<InputFeatures component="ColorInput" element="input" />`
- `<InputSections component="ColorInput" />`
- `<Polymorphic defaultElement="button" changeToElement="a" component="ActionIcon" withNext />`
- `<ServerComponentsIncompatible component="CopyButton" />`
- `<TargetComponent component="HoverCard" />`
- `<WrapperProps component="Radio" />`

Content of these components is usually important for understanding the component. It is better to include the content of those in llm.txt. https://github.com/mantinedev/mantine/tree/master/apps/mantine.dev/src/components/MdxProvider/MdxSharedContent

brookton commented:
Thanks for your help with this @rtivital! I'm a huge fan of Mantine and glad to be contributing.

I believe I've addressed your concerns about the shared content components. 
Instead of removing them, the scripts now:

- Preserve the original JSX examples 
  - Components like `<Polymorphic defaultElement="button" changeToElement="a" component="ActionIcon" withNext />` are kept and real usage with specific props are included
- Include the educational content 
  - Each component's documentation is rendered below its JSX example (e.g., `AutoContrast` explains the color contrast feature, `GetElementRef` shows TypeScript ref usage, etc.)
-  Handle all component props properly 
   - The content adapts based on props like `withVariant`, `refType`, element, etc. to match what the React components would render

  The output now looks like:
```md
  <Polymorphic defaultElement="button" changeToElement="a" component="ActionIcon" withNext />

  ## Polymorphic component

  ActionIcon is a polymorphic component – its default root element is button, but it can be changed to any other element or component
  with component prop:
  [... rest of content ...]
```
  
Hopefully this addresses your comment. Let me know if you'd like any adjustments!

rtivital commented:
Looking better already, several other fixes left:
1. Remove changelogs `### Changelog810 ...` – there is duplicateinformation
2. Metadata leftovers from help.mantine.dev articles

```
export const meta = {
  title:
    'Why can I not use value/label data structure with Autocomplete/TagsInput?',
  description:
    'Learn the difference between Autocomplete and Select components',
  slug: 'autocomplete-value-label',
  category: 'forms',
  tags: [
    'autocomplete',
    'tags-input',
    'data',
    'select',
    'multi-select',
    'MultiSelect',
  ],
  created_at: 'July 15, 2024',
  last_updated_at: 'July 15, 2024',
};
```

3. Some help.mantine.dev articles are broken. They also do not include demo code. See https://raw.githubusercontent.com/mantinedev/mantine/refs/heads/master/apps/help.mantine.dev/src/pages/q/align-input-button.mdx for reference.
```
--------------------------------------------------------------------------------

## FREQUENTLY ASKED QUESTIONS


AlignInputButton,
AlignInputButtonCorrect,
AlignInputButtonStyles,
} from '@/demos/AlignInputButton.demo';

export const meta = {
  title: 'How to align input with a button in a flex container?',
  description:
    'Learn how to align Mantine inputs with buttons using Group component',
  slug: 'align-input-button',
  category: 'styles',
  tags: ['inputs', 'button', 'flexbox', 'centering'],
  created_at: 'February 15, 2024',
  last_updated_at: 'February 15, 2024',
};

If you try to align one of Mantine inputs in a flex container with a button,
you will notice that input is not aligned with the button. This happens because
Mantine inputs have associated elements: label, description and error message.



To align input with a button, you can either use `inputContainer` prop to wrap the button
next to the in a flex container:



Or change error/description position to absolute with [Styles API](https://mantine.dev/styles/styles-api/):




----------------------------------------
```

brookton commented:
Thanks for the feedback @rtivital! Glad we're getting closer.

Just to confirm you want all changelog information ignored for the scripts? 
I've pushed that change to a commit but want to confirm incase you'd expect any unique information living in them that you'd want me to add logic to find and include or something.

Chipping away at the rest now as well!

rtivital commented:
Great, final changes left, please do the following:
- Fix reported CI errors (`npm test` locally to validate)
- Remove LLM.md from the repository (it was needed only for the review)
- Add llms.txt to `.gitignore` (the file must not be shipped with the repository)

brookton commented:
Amazing!

Thanks again for your help with this @rtivital. It's been a pleasure.

Completed all those changes, should be finalized now! 🎉🥳🎊

brookton commented:
Hey @rtivital 

I noticed some additional issues and pushed some important changes.

  - Added complete extraction of CSS selectors, CSS variables, and data attributes from styles-api
  files
  - Improved template literal parsing to handle complex nested expressions (e.g.,
  ${Object.keys().map()})
  - Fixed variable interpolation in multi-file demos (e.g., ${tsxCode}, ${cssCode})
  - Handles escaped characters properly (\$ → $, \`` → ``  ``)
  - Fixed issue where import statements in demo code were being incorrectly removed
  - Added proper style import comments for all packages that require them
  - Fixed package name resolution (e.g., @mantine/code-highlight instead of @mantine/x)

I outputted the LLM.md again for you to compare if you'd like. But I believe this to be a comprehensive generative script now. If you notice any other errors in that output though please let me know.

Thanks so much again,

David

rtivital commented:
Thanks!
